src/Microsoft.VisualStudio.Threading/AsyncReaderWriterLock.cs

```diff
@@ -1600,6 +1664,37 @@ private bool TryInvokeAllReadersIfAppropriate(bool searchAllWaiters)
             return invoked;
         }
 
+        private void TryInvokeAllDependentReadersIfAppropriate()
+        {
+            lock (this.syncObject)
+            {
+                if (this.issuedWriteLocks.Count == 0 && this.waitingWriters.Count > 0 && this.waitingReaders.Count > 0 && (this.issuedReadLocks.Count > 0 || this.issuedUpgradeableReadLocks.Count > 0))
+                {
+                    HashSet<JoinableTask>? dependentTasks = JoinableTaskDependencyGraph.GetDependentTasksFromCandidates(
+                        this.issuedReadLocks.Concat(this.issuedUpgradeableReadLocks).Where(w => w.AmbientJoinableTask != null).Select(w => w.AmbientJoinableTask!),
+                        this.waitingReaders.Where(w => w.AmbientJoinableTask != null).Select(w => w.AmbientJoinableTask!));
+
+                    if (dependentTasks.Count > 0)
+                    {
+                        int pendingCount = this.waitingReaders.Count;
+                        while (pendingCount-- != 0)
+                        {
+                            Awaiter? pendingReader = this.waitingReaders.Dequeue();
+                            JoinableTask? readerContext = pendingReader?.AmbientJoinableTask;
+                            if (readerContext != null && dependentTasks.Contains(readerContext))
+                            {
+                                this.IssueAndExecute(pendingReader!);
+                            }
+                            else
+                            {
+                                this.waitingReaders.Enqueue(pendingReader!);
```

Comment:
It seems odd that you're evidently prepared for `pendingReader is null` yet you shove it (back) into a collection that is evidently typed to guarantee there are no null elements. I feel like this code both doesn't trust the null ref annotations and doesn't mind violating them either. Can this be changed to depend on them more? Except when checking inputs at public entrypoints, we should be able to trust and honor them.